C# 6.0 Feature: using static

BillWagner opened the pull request:
Current spec text for using static.

~~Creating a draft PR because of outstanding TODO item.~~ TODO item is complete.

Fixes #51
Fixes #199

BillWagner committed:
updates for using static

Current spec text for using static

proofread fixes

BillWagner committed:
add example for inheritance and static using

I chose to use non-descript names rather than provide an example from the class library. I think that's sufficient to understand, and doesn't beg more questions about the design decision (even in an informative example).

BillWagner committed:
fix duplicate header

BillWagner committed:
fix CI build

BillWagner committed:
fix build warning

jskeet commented on standard/namespaces.md:
```diff
@@ -57,7 +57,7 @@ A *namespace_declaration* may occur as a top-level declaration in a *compilation
 Namespaces are implicitly `public` and the declaration of a namespace cannot include any access modifiers.
 
-Within a *namespace_body*, the optional *using_directive*s import the names of other namespaces and types, allowing them to be referenced directly instead of through qualified names. The optional *namespace_member_declaration*s contribute members to the declaration space of the namespace.
+Within a *namespace_body*, the optional *using_directive*s import the names of other namespaces, types and members, allowing them to be referenced directly instead of through qualified names. The optional *namespace_member_declarations* contribute members to the declaration space of the namespace. Note that all *using_directive*s must appear before any member declarations.
```
Any reason *namespace_member_declaration*s was changed to *namespace_member_declarations*? I think the former is correct. (The difference is in whether the trailing "s" is in italics.)

jskeet commented on standard/namespaces.md:
```diff
@@ -133,14 +133,17 @@ The alias introduced by an *extern_alias_directive* is very similar to the alias
-The scope of a *using_directive* extends over the *namespace_member_declaration*s of its immediately containing compilation unit or namespace body. The scope of a *using_directive* specifically does not include its peer *using_directive*s. Thus, peer *using_directive*s do not affect each other, and the order in which they are written is insignificant. In contrast, the scope of an *extern_alias_di ...
+A *using_static_directive* (§using-static-directives) imports the nested types and static members of a type.
+
+The scope of a *using_directive* extends over the *namespace_member_declarations* of its immediately containing compilation unit or namespace body. The scope of a *using_directive* specifically does not include its peer *using_directive*s. Thus, peer *using_directive*s do not affect each other, and the order in which they are written is insignificant. In contrast, the scope of an *extern_alias_di ...
```
Ditto (trailing s) for *using_directives* in the final sentence. (I suspect then that's changed, this paragraph will be as it was before, and we'll only have the new paragraph before it.)

jskeet commented on standard/namespaces.md:
```diff
@@ -446,10 +449,130 @@ Because names may be ambiguous when more than one imported namespace introduces
+>     using static C;
+>     class B
+>     {
+>         void M() 
```
We normally use brace-at-end-of-line for method declarations. (I know, I know...)

jskeet commented on standard/namespaces.md:
```diff
@@ -446,10 +449,130 @@ Because names may be ambiguous when more than one imported namespace introduces
+> {
+>     class A {}
+> }
+> class C
```
A blank line between each top-level declaration? This is a bit hard to read as-is. (We might want to talk about this - it looks like a lot of examples don't have lines between top-level declarations.)

BillWagner replied:
I did add them here.
I didn't resolve the conversation so we can discuss at the upcoming meeting.

jskeet replied:
Group decided to add a blank line between peer declarations (other than obvious cases where it would harm readability, e.g. where there are 5 methods one after another, all with an implementation of `{}`, just for signatures etc).

@BillWagner: Do you want to make the change consistently in this PR and we'll work out what to do elsewhere?

BillWagner replied:
I'll make the changes consistently in the areas touched by this PR.

I'll keep notes and make an open issue for all formatting so we have a record of our preferred formatting. Maybe we can get some community help on that task.

jskeet commented on standard/namespaces.md:
```diff
@@ -446,10 +449,130 @@ Because names may be ambiguous when more than one imported namespace introduces
+>    class A 
+>    {
+>         public class B{}
+>         public static B M(){ return new B(); }
```
Can we use an expression-bodied method for this? If not, can we have a space between `()` and `{`?

jskeet commented on standard/namespaces.md:
````diff
@@ -446,10 +449,130 @@ Because names may be ambiguous when more than one imported namespace introduces
+>     }
+> }
+> ```
+> Above, within member declarations in the `N2` namespace, the static members and nested types of `N1.A` are directly available, and thus the method `N` is able to reference both the `B` and `M` members of `N1.A`. *end example*
````
```suggestion
> In the preceding code, within member declarations in the `N2` namespace, the static members and nested types of `N1.A` are directly available, and thus the method `N` is able to reference both the `B` and `M` members of `N1.A`. *end example*
```

(Or something similar - just "Above" sounds a bit bare.)

jskeet commented on standard/namespaces.md:
```diff
@@ -446,10 +449,130 @@ Because names may be ambiguous when more than one imported namespace introduces
+> namespace N2
+> {
+>     using static N1.A;
+>         class B
```
Outdent these 9 lines?

jskeet commented on standard/namespaces.md:
```diff
@@ -446,10 +449,130 @@ Because names may be ambiguous when more than one imported namespace introduces
+>     using static N1.B;
+>         class C
+>         {
+>             void N() 
```
Brace at end of line.

jskeet commented on standard/namespaces.md:
```diff
@@ -446,10 +449,130 @@ Because names may be ambiguous when more than one imported namespace introduces
+>     using static N1.A;
+>         class B
+>         {
+>             void N() 
```
Brace at end of line

jskeet reviewed (approved):
Broadly fine - just some formatting nits.

RexJaeschke reviewed (approved):
I have nothing more to add.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Jon Skeet <skeet@pobox.com>

BillWagner committed:
respond to feedback

BillWagner committed:
braces on end of line

I wanted to check other clauses in the standard to compare. Jon is correct.

KalleOlaviNiemitalo commented on standard/namespaces.md:
````diff
@@ -446,10 +449,129 @@ Because names may be ambiguous when more than one imported namespace introduces
+>         }
+> }
+> ```
+> the *using_static_directive* imports the method `M2` contained in `N1.B`, but does not import the method `M` contains in `N1.A`. Thus, the reference to `M` in the body of `C.N` results in a compile-time error because no members named `M` are in scope. Developers must add a second `using static` directive to specify that the methods in `N1.A` should also be imported. *end example*
````
method `M` contain**ed** in `N1.A`.

KalleOlaviNiemitalo commented on standard/namespaces.md:
```diff
@@ -446,10 +449,129 @@ Because names may be ambiguous when more than one imported namespace introduces
+> }
+> namespace N2
+> {
+>     using static N1.B;
+>         class C
```
The indentation looks strange.

BillWagner committed:
respond to public feedback.

jskeet commented:
Meeting is happy to merge when formatting is sorted.

BillWagner committed:
Format samples

Make the following changes to the code samples in this clause:

- Add a blank line between peer elements *except* for multiple `using` directives and multiple `alias` directives. (I did add a blank line between `using` and `alias` directives.
- Use Allman brace style on non-empty `class` declarations. (I'd prefer using Allman brace style on methods, but the standard currently uses K&R style for methods consistently.

BillWagner commented on standard/namespaces.md:
```diff
@@ -669,7 +713,9 @@ Using this notation, the meaning of a *qualified_alias_member* is determined as
 > class A { }
-> class C {
+>
+> class C
```
This was one of 2 instances for K&R braces on `class` declarations. That's why I changed this one.

BillWagner commented on standard/namespaces.md:
```diff
@@ -680,8 +726,11 @@ Using this notation, the meaning of a *qualified_alias_member* is determined as
-> class C {
+>
+> class C 
+> {
```
second instance of K&R braces on a `class`

BillWagner commented on standard/namespaces.md:
```diff
@@ -166,6 +167,7 @@ Within global attributes and member declarations in a compilation unit or namesp
 > namespace N3
 > {
 >     using A = N1.N2.A;
+>
```
@jskeet This was one type of change I wanted to check on. I think it's better by having the blank line, but it's a very mild preferences. 

For our "coding style", I'd like to say different elements should have a blank line between them. On the other hand, I like a rule that doesn't add a blank line where the element is an empty class or namespace.

See also line 179, 192, 201-203.

jskeet replied:
Mmm, interesting. I probably wouldn't have added blank lines in 179, 192, 201-203... but I'm quite happy for us to do so, and that should probably be the default. We should just allow ourselves to have an exception for things like:

```csharp
public void M(int x) {}
public void M(string y) {}
public void M(int x, string y = null) {}
```

... in an example about overload resolution, for example.

BillWagner replied:
I agree. I think that's easily handled by adding methods to my exception in the previous comment.

BillWagner replied:
I opened #358 to start the coding standard discussion. 

I'll merge this.

BillWagner reviewed (commented):
@jskeet I tried to make the changes in this clause to develop our general practice. 

I followed the notes from the meeting, and then applied my own opinion on whitespace, because I was doing the work. 😀
